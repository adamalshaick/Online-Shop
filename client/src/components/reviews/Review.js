import React from "react";
import PropTypes from "prop-types";
import { CardContent, Card, Typography } from "@material-ui/core";

const displayStars = review => {
  const stars = [];
  for (let i = 0; i < review.rate; i++) {
    stars.push(<i style={{ color: "gold" }} className="fas fa-star" />);
  }
  return stars;
};

export const Review = ({ review }) => {
  return (
    <div className="col-md-6 mt-3 entry">
      <Card>
        <CardContent>
          <Typography color="textSecondary">
            rate: <span>{displayStars(review)}</span>
          </Typography>
          <Typography color="textSecondary">
            name: <strong>{review.buyer}</strong>
          </Typography>
          <Typography color="textSecondary">
            review: <strong>{review.text}</strong>
          </Typography>
        </CardContent>
      </Card>
    </div>
  );
};

Review.propTypes = {
  review: PropTypes.object.isRequired
};

export default Review;
