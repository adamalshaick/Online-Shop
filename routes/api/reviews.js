const express = require("express");
const mongoose = require("mongoose");
const passport = require("passport");
const router = express.Router();

// Review model
const Review = require("../../models/Review");

// Validation
const validateReviewInput = require("../../validation/review");

// @route POST api/review/:id
// @desc  Add review
// @access Private
router.post(
  "/:id",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    const { errors, isValid } = validateReviewInput(req.body);
    // Check Validation
    if (!isValid) {
      // If any errors, send 400 with errors object
      return res.status(400).json(errors);
    }
    const newReview = new Review({
      text: req.body.text,
      rate: req.body.rate,
      seller: req.params.id,
      buyer: req.body.name
    });
    newReview.save().then(review => res.json(review));
  }
);

// @route GET api/reviews
// @desc  Get reviews
// @access Private
router.get(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    Review.find()
      .then(reviews => res.json(reviews))
      .catch(err =>
        res.status(404).json({ noreviewsfound: "No reviews found" })
      );
  }
);

module.exports = router;
