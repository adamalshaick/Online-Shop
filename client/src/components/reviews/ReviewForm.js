import React from "react";
import PropTypes from "prop-types";
import TextAreaFieldGroup from "../common/TextAreaFieldGroup";
import NumberInput from "../common/NumberInput";
import { ReviewCard } from "../common/styles/Review";
import { ShadowCard } from "../common/styles/ShadowCard";

export const ReviewForm = ({
  onClickRev,
  errors,
  text,
  rate,
  onSubmit,
  onChange
}) => {
  return (
    <>
      <div className="row">
        <ReviewCard className="col-md-6 col-lg-4">
          <ShadowCard className="card p-0">
            <div className="card-header text-right">
              <span
                style={{ fontSize: "1.3rem", fontWeight: "bold" }}
                className="float-left"
              >
                Write a review
              </span>
              <button onClick={onClickRev} className="btn btn-danger">
                <i className="fas fa-times" />
              </button>
            </div>
            <div className="card-body">
              <form onSubmit={onSubmit}>
                <div className="form-group">
                  <div
                    className="float-right mb-3"
                    style={{
                      display: "flex",
                      fontSize: "1.2rem",
                      fontWeight: "bold"
                    }}
                  >
                    Add a Rating:
                    <NumberInput
                      name="rate"
                      value={rate}
                      error={errors.rate}
                      type="text"
                      onChange={onChange}
                    />
                    /5
                  </div>
                </div>
                <TextAreaFieldGroup
                  name="text"
                  value={text}
                  error={errors.text}
                  type="text"
                  onChange={onChange}
                />
                <button type="submit" className="btn btn-dark float-right">
                  Submit
                </button>
              </form>
            </div>
          </ShadowCard>
        </ReviewCard>
      </div>
    </>
  );
};

ReviewForm.propTypes = {
  onClickRev: PropTypes.func.isRequired,
  errors: PropTypes.object,
  text: PropTypes.string.isRequired,
  rate: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
};

export default ReviewForm;
