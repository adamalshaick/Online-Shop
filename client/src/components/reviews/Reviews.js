import React from "react";
import PropTypes from "prop-types";
import Review from "./Review";
import fetchReviews from "../common/hoc/fetchReviews";

const Reviews = ({ user, review }) => {
  let reviewItems;
  if (review.reviews.length > 0) {
    const usersReview = review.reviews.filter(
      review => review.seller === user._id
    );
    if (usersReview.length > 0) {
      reviewItems = usersReview.map(review => (
        <Review key={review._id} review={review} />
      ));
    } else {
      reviewItems = (
        <div className="text-center col-12">
          <p>
            <i>You don't have any reviews yet</i>
          </p>
        </div>
      );
    }
  } else {
    reviewItems = (
      <div className="text-center">
        <p>
          <i>User has no reviews yet</i>
        </p>
      </div>
    );
  }
  return (
    <div className="container">
      <div className="row">{reviewItems}</div>
    </div>
  );
};

Reviews.propTypes = {
  review: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired
};

export default fetchReviews(Reviews);
