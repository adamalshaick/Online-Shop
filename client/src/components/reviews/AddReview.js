import React, { Component } from "react";
import { connect } from "react-redux";
import { addReview } from "../../actions/reviewActions";
import ReviewForm from "../reviews/ReviewForm";
import { Button } from "../common/styles/Button";
import handleInputErrors from "../common/hoc/handleInputErrors";
import PropTypes from "prop-types";

export class AddReview extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showReviewInput: false,
      text: "",
      rate: ""
    };
  }

  onClickRev = () => {
    this.setState(prevState => ({
      showReviewInput: !prevState.showReviewInput
    }));
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onSubmit = (id, name) => e => {
    e.preventDefault();
    const reviewData = {
      text: this.state.text,
      rate: this.state.rate,
      id: id,
      name: name
    };
    this.props.addReview(reviewData);
  };
  render() {
    const { errors, user } = this.props;
    return (
      <>
        <Button
          id="toggleBtn"
          className="btn btn-light btn-lg mt-2 mb-3"
          onClick={this.onClickRev}
        >
          Write a review
        </Button>
        <div>
          {this.state.showReviewInput ? (
            <ReviewForm
              id="reviewForm"
              onClickRev={this.onClickRev}
              text={this.state.text}
              rate={this.state.rate}
              errors={errors}
              onChange={this.onChange}
              onSubmit={this.onSubmit(user._id, user.name)}
            />
          ) : null}
        </div>
      </>
    );
  }
}

AddReview.propTypes = {
  addReview: PropTypes.func.isRequired,
  errors: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired
};

export default connect(
  null,
  { addReview }
)(handleInputErrors(AddReview));
