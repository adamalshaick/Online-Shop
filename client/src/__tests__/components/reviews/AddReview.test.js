import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { AddReview } from "../../../components/reviews/AddReview";

const mockAddReview = jest.fn();

const Props = {
  user: {},
  errors: {}
};

const wrapper = shallow(<AddReview {...Props} addReview={mockAddReview} />);
describe("render component", () => {
  it("renders component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("toggles btn", () => {
  wrapper.find("#toggleBtn").simulate("click");
  expect(wrapper.find("#reviewForm").length).toEqual(1);
  wrapper.find("#toggleBtn").simulate("click");
  expect(wrapper.find("#reviewForm").length).toEqual(0);
});

describe("calls mock add review fn", () => {
  wrapper.find("#toggleBtn").simulate("click");
  wrapper.find("#reviewForm").simulate("submit", { preventDefault() {} });
  expect(mockAddReview.mock.calls.length).toEqual(1);
});

// describe("reviews", () => {
//   it("adds review", () => {
//     wrapper.find("#addReview").simulate("click");

//   });
// });

// import React, { Component } from "react";
// import { connect } from "react-redux";
// import { addReview } from "../../actions/reviewActions";
// import ReviewForm from "../reviews/ReviewForm";
// import { Button } from "../common/styles/Button";
// import handleInputErrors from "../common/hoc/handleInputErrors";

// class AddReview extends Component {
//   constructor(props) {
//     super(props);
//     this.state = {
//       showReviewInput: false,
//       text: "",
//       rate: ""
//     };
//   }

//   onClickRev = () => {
//     this.setState(prevState => ({
//       showReviewInput: !prevState.showReviewInput
//     }));
//   };

//   onChange = e => {
//     this.setState({ [e.target.name]: e.target.value });
//   };

//   onSubmit = (id, name) => e => {
//     e.preventDefault();
//     const reviewData = {
//       text: this.state.text,
//       rate: this.state.rate,
//       id: id,
//       name: name
//     };
//     this.props.addReview(reviewData);
//   };
//   render() {
//     const { errors, user } = this.props;
//     return (
//       <>
//         <Button
//           className="btn btn-light btn-lg mt-2 mb-3"
//           onClick={this.onClickRev}
//         >
//           Write a review
//         </Button>
//         <div>
//           {this.state.showReviewInput ? (
//             <ReviewForm
//               onClickRev={this.onClickRev}
//               text={this.state.text}
//               rate={this.state.rate}
//               errors={errors}
//               onChange={this.onChange}
//               onSubmit={this.onSubmit(user._id, user.name)}
//             />
//           ) : null}
//         </div>
//       </>
//     );
//   }
// }

// export default connect(
//   null,
//   { addReview }
// )(handleInputErrors(AddReview));
