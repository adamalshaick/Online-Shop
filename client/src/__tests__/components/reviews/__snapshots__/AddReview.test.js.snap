// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<Fragment>
  <styled.button
    className="btn btn-light btn-lg mt-2 mb-3"
    id="toggleBtn"
    onClick={[Function]}
  >
    Write a review
  </styled.button>
  <div>
    <ReviewForm
      errors={Object {}}
      id="reviewForm"
      onChange={[Function]}
      onClickRev={[Function]}
      onSubmit={[Function]}
      rate=""
      text=""
    />
  </div>
</Fragment>
`;
